src/lem/constrainer.rs

```diff
@@ -125,56 +381,341 @@ impl LEM {
+                        bail!("{} not allocated", preimg[0].name());
                     };
-                    let z_ptr = {
-                        if Self::on_concrete_path(&branch_path_info)? {
-                            let Some(ptr) = witness.ptrs.get(tgt.name()) else {
-                                bail!("Couldn't retrieve witness {}", tgt.name());
-                            };
-                            store.hydrate_ptr(ptr)?
-                        } else {
-                            ZPtr::dummy()
-                        }
+                    let Some(i2) = alloc_ptrs.get(preimg[1].name()) else {
+                        bail!("{} not allocated", preimg[1].name());
+                    };
+
+                    // Allocate new pointer containing expected hash value
+                    let alloc_hash = Self::allocate_ptr(
+                        cs,
+                        &Self::z_ptr_from_valuation(
+                            &branch_path_info,
+                            valuation,
+                            hash.name(),
+                            store,
+                        )?,
+                        hash.name(),
+                        &alloc_ptrs,
+                    )?;
+
+                    // Stack expected hash, preimage and tag, together with virtual path information,
+                    // such that only concrete path hashes are indeed calculated in the next available
+                    // hash slot.
+                    Self::stack_hash_slots(
+                        &branch_path_info,
+                        &mut hash_slots,
+                        &slots,
+                        hash.clone(),
+                        AllocHashPreimage::A2(i1.clone(), i2.clone()),
+                        Some(*tag),
+                    )?;
+
+                    // Insert hash value pointer in the HashMap
+                    alloc_ptrs.insert(hash.name(), alloc_hash.clone());
+                }
+                LEMOP::Unhash2(preimg, hash) => {
+                    // Get preimage from allocated pointers
+                    let i1 = Self::allocate_ptr(
+                        cs,
+                        &Self::z_ptr_from_valuation(
+                            &branch_path_info,
+                            valuation,
+                            preimg[0].name(),
+                            store,
+                        )?,
+                        preimg[0].name(),
+                        &alloc_ptrs,
+                    )?;
+
+                    let i2 = Self::allocate_ptr(
+                        cs,
+                        &Self::z_ptr_from_valuation(
+                            &branch_path_info,
+                            valuation,
+                            preimg[1].name(),
+                            store,
+                        )?,
+                        preimg[1].name(),
+                        &alloc_ptrs,
+                    )?;
```

Comment:
```rust
                    let i = preimg.iter().map(|pre| {
                        let name = pre.name();
                        let res_ptr = 
                            Self::z_ptr_from_valuation(
                                &branch_path_info,
                                valuation,
                                name,
                                store,
                            );
                        res_ptr.and_then(|ref ptr| {
                            Self::allocate_ptr(
                                cs,
                                ptr,
                                name,
                                &alloc_ptrs,
                            )
                        })
                    }).collect::<Result<Vec<_>>>()?;
```

... and then use `i[0]`, `i[1]` ... more generally large chunks can be rewritten to used a vectorized style, and hence parallelize / avoid errors.